generic/VolumetricMaxUnpooling.c

```diff
@@ -0,0 +1,272 @@
+{
+  THTensor *input = luaT_checkudata(L, 2, torch_Tensor);
+  THTensor *indices = luaT_getfieldcheckudata(L, 1, "indices", torch_Tensor);
+  THTensor *output = luaT_getfieldcheckudata(L, 1, "output", torch_Tensor);
+  int otime = luaT_getfieldcheckint(L, 1, "otime");
+  int owidth = luaT_getfieldcheckint(L, 1, "owidth");
+  int oheight = luaT_getfieldcheckint(L, 1, "oheight");
+  int dT = luaT_getfieldcheckint(L, 1, "dT");
+  int dH = luaT_getfieldcheckint(L, 1, "dH");
+  int dW = luaT_getfieldcheckint(L, 1, "dW");
+  int dimw = 3;
+  int dimh = 2;
+  int dimt = 1;
+  int nbatch = 1;
+  int nslices;
+  int itime;
+  int iheight;
+  int iwidth;
+  real *input_data;
+  real *output_data;
+  real *indices_data;
+
+  luaL_argcheck(L, input->nDimension == 4 || input->nDimension == 5 , 2, "4D or 5D (batch mode) tensor expected");
+  if (!THTensor_(isSameSizeAs)(input, indices)){
+    THError("Invalid input size w.r.t current indices size");
+  }  
+
+  if (input->nDimension == 5) 
+  {
+    nbatch = input->size[0];
+    dimt++;
+    dimw++;
+    dimh++;
+  }
+
+  /* sizes */
+  nslices = input->size[dimt-1];
+  itime = input->size[dimt];
+  iheight = input->size[dimh];
+  iwidth = input->size[dimw];
+
+  /* get contiguous input */
+  input = THTensor_(newContiguous)(input);
+  indices = THTensor_(newContiguous)(indices);
+
+  /* resize output */
+  if (input->nDimension == 4)
+  {
+    THTensor_(resize4d)(output, nslices, otime, oheight, owidth);
+    THTensor_(zero)(output);
+
+    input_data = THTensor_(data)(input);
+    output_data = THTensor_(data)(output);
+    indices_data = THTensor_(data)(indices);
+
+    nn_(VolumetricMaxUnpooling_updateOutput_frame)(input_data, output_data,
+                                              indices_data,
+                                              nslices,
+                                              itime, iwidth, iheight,
+                                              otime, owidth, oheight,
+                                              dT, dW, dH);
+  }
+  else
+  {
+    long p;
+
+    THTensor_(resize5d)(output, nbatch, nslices, otime, oheight, owidth);
+    THTensor_(zero)(output);
+
+    input_data = THTensor_(data)(input);
+    output_data = THTensor_(data)(output);
+    indices_data = THTensor_(data)(indices);
+
+#pragma omp parallel for private(p)
+    for (p = 0; p < nbatch; p++)
+    {
+      nn_(VolumetricMaxUnpooling_updateOutput_frame)(input_data+p*nslices*itime*iwidth*iheight, output_data+p*nslices*otime*owidth*oheight,
+                                                indices_data+p*nslices*itime*iwidth*iheight,
+                                                nslices,
+                                                itime, iwidth, iheight,
+                                                otime, owidth, oheight,
+                                                dT, dW, dH);
+    }
+  }
+
+  /* cleanup */
+  THTensor_(free)(input);
```

Comment:
Indeed there is a leak. The leak is also present in the spatial version of this module, which should be fixed. cc @clement-masson 
There is no "real" leak in fact because `indices` is contiguous and kept forever.